src/sanityImage.test.ts

```diff
@@ -0,0 +1,294 @@
+        .grab({
+          name: q.string(),
+          cover: q.sanityImage("cover", {
+            withAsset: ["base"],
+          }),
+        })
+    );
+
+    expect(query).toBe(
+      `*[_type == 'pokemon'][0..1]{name, "cover": cover{_key, _type, "asset": asset->{_id, _type, _rev, extension, mimeType, originalFilename, path, sha1hash, size, url, _updatedAt}}}`
+    );
+    const im = data?.[0]?.cover;
+    invariant(im);
+    expect(im.asset._id === "image-1-jpg").toBeTruthy();
+    expect(im.asset._type === "sanity.imageAsset").toBeTruthy();
+    expect(im.asset._rev === "X6HgJNl2Cktkcl6TQwg3gv").toBeTruthy();
+    expect(im.asset.extension === "jpg").toBeTruthy();
+    expect(im.asset.mimeType === "image/jpeg").toBeTruthy();
+    expect(im.asset.originalFilename === "pokemon-1.jpg").toBeTruthy();
+    expect(
+      im.asset.path ===
+        "images/nfttuagc/production/ed158069c3b44124a310d7a107998e06bf12e90e-1000x500.jpg"
+    ).toBeTruthy();
+    expect(
+      im.asset.sha1hash === "ed158069c3b44124a310d7a107998e06bf12e90e"
+    ).toBeTruthy();
+    expect(im.asset.size === 37594).toBeTruthy();
+    expect(
+      im.asset.url ===
+        "https://cdn.sanity.io/images/nfttuagc/production/ed158069c3b44124a310d7a107998e06bf12e90e-1000x500.jpg"
+    ).toBeTruthy();
+    expect(im.asset._updatedAt instanceof Date).toBeTruthy();
+  });
+
+  it("can query fetch image asset data with dimensions metadata", async () => {
+    const { query, data } = await runPokemonQuery(
+      q("*")
+        .filter("_type == 'pokemon'")
+        .slice(0, 1)
+        .grab({
+          name: q.string(),
+          cover: q.sanityImage("cover", {
+            withAsset: ["dimensions"],
+          }),
+        })
+    );
+
+    expect(query).toBe(
+      `*[_type == 'pokemon'][0..1]{name, "cover": cover{_key, _type, "asset": asset->{"metadata": metadata{"dimensions": dimensions{_type, aspectRatio, height, width}}}}}`
+    );
+    const dimensions = data?.[0]?.cover?.asset?.metadata?.dimensions;
+    invariant(dimensions);
+    expect(dimensions._type === "sanity.imageDimensions").toBeTruthy();
+    expect(dimensions.aspectRatio === 2).toBeTruthy();
+    expect(dimensions.height === 500).toBeTruthy();
+    expect(dimensions.width === 1000).toBeTruthy();
+  });
+
+  it("can query fetch image asset data with location metadata", async () => {
+    const { query, data } = await runPokemonQuery(
+      q("*")
+        .filter("_type == 'pokemon'")
+        .slice(0, 1)
+        .grab({
+          name: q.string(),
+          cover: q.sanityImage("cover", {
+            withAsset: ["location"],
+          }),
+        })
+    );
+
+    expect(query).toBe(
+      `*[_type == 'pokemon'][0..1]{name, "cover": cover{_key, _type, "asset": asset->{"metadata": metadata{"location": location{_type, lat, lng}}}}}`
+    );
+    const location = data?.[0]?.cover?.asset?.metadata?.location;
+    invariant(location);
+    expect(location._type === "geopoint").toBeTruthy();
```

Comment:
```suggestion
    expect(location._type).toEqual("geopoint");
```
Maybe restructure the expect calls to be like this, for better error messages.

Reply:
I've been dying to have a conversation with someone about this! The test suite here looks _different_ than how I'd typically write unit tests, primarily to test the TS types! The unit tests serve a dual purpose: runtime behavior, and TS typings.

In this scenario, `location._type` should be the exact literal `"geopoint"` – and it's hard to capture that sort of type-checking with tests like `expect(location._type).toEqual("geopoint")`, since that latter example doesn't make any type-level assertions about what the type/literal value of `_type` is.

I'm still trying to flesh my thoughts out regarding this, but that's a little insight into why these tests look the way they do.

Reply:
That makes sense. It's fine as-is, then.